casepropods/family_connect_subscription/plugin.py

```diff
@@ -78,6 +112,69 @@ def format_schedule(self, schedule):
             schedule['month_of_year'], schedule['day_of_week'])
         return prettify_cron(cron_schedule)
 
+    def cancel_subscriptions(self, subscription_ids):
+        params = {'active': False}
+        for subscription in subscription_ids:
+            try:
+                self.stage_based_messaging_api.update_subscription(
+                    subscription, params)
+            except HTTPServiceError:
+                return False
+        return True
+
+    def perform_action(self, type_, params):
+        if type_ == "cancel_subs":
+            subscription_ids = params.get("subscription_ids", [])
+            success = self.cancel_subscriptions(subscription_ids)
+            if not success:
+                return (False,
+                        {"message": "Failed to cancel some subscriptions"})
+            return (True, {"message": "cancelled all subscriptions"})
+
+        if type_ == "full_opt_out":
+            opted_out = False
+            subs_cancelled = False
+
+            # Create OptOut
+            contact_id = params["contact_id"]
+            contact = Contact.objects.get(pk=contact_id)
+            identity = contact.uuid
+
+            opt_out_url = settings.IDENTITY_API_ROOT + "api/v1/optout/"
+            identity_token = settings.IDENTITY_AUTH_TOKEN
+            headers = {
+                'Authorization': "Token " + identity_token,
+                'Content-Type': "application/json"
+            }
+
+            # Opt-outs have to have an address. This kinda sucks though
+            if contact.urns:
+                addr_type, address = contact.urns[0].split(':', 1)
+                response = requests.post(
+                    opt_out_url, headers=headers,
+                    json={'identity': identity, 'optout_type': "forget",
+                          'address_type': addr_type, 'address': address,
+                          'request_source': 'casepro'},
+                )
+                if response.status_code == 201:
+                    opted_out = True
+
+            # Cancel Subscriptions
+            subscription_ids = params.get("subscription_ids", [])
+            subs_cancelled = self.cancel_subscriptions(subscription_ids)
```

Comment:
Again, maybe we can split out the 'create optout' and the 'cancel subscriptions' into functions that return `True` or `False`, and then just call them, and handle the results as we already have below.